backend/ml/mehestan/online_heuristics.py

```diff
@@ -0,0 +1,519 @@
+            & (df_all_comparison_of_user_for_criteria.entity_b == entity_id_a)
+        ].empty
+    ):
+        logger.warning(
+            "_run_online_heuristics_for_criterion :  \
+            no comparison found for '%s' with '%s' and criteria '%s'",
+            entity_id_a,
+            entity_id_b,
+            criteria,
+        )
+        return False
+    return True
+
+
+def run_online_heuristics(
+    ml_input: MlInput,
+    uid_a: str,
+    uid_b: str,
+    user_id: str,
+    poll: Poll,
+    delete_comparison_case: bool,
+    parallel_computing: bool = True,
+):
+    """
+    This function use multiprocessing.
+
+        1. Always close all database connections in the main process before
+           creating forks. Django will automatically re-create new database
+           connections when needed.
+
+        2. Do not pass Django model's instances as arguments to the function
+           run by child processes. Using such instances in child processes
+           will raise an exception: connection already closed.
+
+        3. Do not fork the main process within a code block managed by
+           a single database transaction.
+
+    See the indications to close the database connections:
+        - https://www.psycopg.org/docs/usage.html#thread-and-process-safety
+        - https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNECT
+
+    See how django handles database connections:
+        - https://docs.djangoproject.com/en/4.0/ref/databases/#connection-management
+    """
+    logger.info("Online Heuristic Mehestan for poll '%s': Start", poll.name)
+
+    # Avoid passing model's instances as arguments to the function run by the
+    # child processes. See this method docstring.
+    poll_pk = poll.pk
+    criteria = poll.criterias_list
+
+    _run_online_heuristics_for_criterion(
+        ml_input=ml_input,
+        poll_pk=poll_pk,
+        criteria=poll.main_criteria,
+        uid_a=uid_a,
+        uid_b=uid_b,
+        user_id=user_id,
+        delete_comparison_case=delete_comparison_case,
+    )
+    save_tournesol_scores(poll)
+    logger.info(
+        "Online Heuristic Mehestan for poll '%s': main_criteria Done", poll.name
+    )
+
+    partial_online_heuristics = partial(
+        _run_online_heuristics_for_criterion,
+        ml_input=ml_input,
+        poll_pk=poll_pk,
+        uid_a=uid_a,
+        uid_b=uid_b,
+        user_id=user_id,
+        delete_comparison_case=delete_comparison_case,
+    )
+    # compute each criterion in parallel
+    remaining_criteria = [c for c in criteria if c != poll.main_criteria]
+    if parallel_computing:
+        os.register_at_fork(before=db.connections.close_all)
+
+        cpu_count = os.cpu_count() or 1
+        with Pool(processes=max(1, cpu_count - 1)) as pool:
+            for _ in pool.imap_unordered(
+                partial_online_heuristics,
+                remaining_criteria,
+            ):
+                pass
```

Comment:
When would we use parellel_computing in this context? Is there a significant gain to expect when running the online heuristics for a single user?
If that's the case, we need to be cautious about the overhead of initializing the threads. At least, the `Pool` could be created once. And as the API is already deployed with multiple workers, the number of processes should be fine-tuned to avoid DoS on the server.

Reply:
I did not success to make the parallel computing working , maybe I should delete that try for the first version to focus the PR on the algorithm rather some technical details and let you investigate if needed in a enhanced version